mirrord-agent/src/runtime.rs

```diff
@@ -1,68 +1,65 @@
 use containerd_client::{
     connect,
-    services::v1::{containers_client::ContainersClient, GetContainerRequest},
+    services::v1::{tasks_client::TasksClient, GetRequest},
     tonic::Request,
     with_namespace,
 };
 use nix::sched::setns;
-use serde::{Deserialize, Serialize};
+pub(crate) enum Runtime {}
 
-const CONTAINERD_SOCK_PATH: &str = "/run/containerd/containerd.sock";
-const DEFAULT_CONTAINERD_NAMESPACE: &str = "k8s.io";
-
-#[derive(Serialize, Deserialize, Debug)]
-struct Namespace {
-    #[serde(rename = "type")]
-    ns_type: String,
-    path: Option<String>,
-}
-
-#[derive(Serialize, Deserialize, Debug)]
-struct LinuxInfo {
-    namespaces: Vec<Namespace>,
-}
-
-#[derive(Serialize, Deserialize, Debug)]
-struct Spec {
-    linux: LinuxInfo,
+impl Runtime {
+    pub async fn get_container_pid(container_id: &str, container_runtime: &str) -> Result<PathBuf> {
+        match container_runtime {
+            "docker" => get_docker_container_pid(container_id.to_string()).await,
+            "containerd" => get_containerd_container_pid(container_id.to_string()).await,
+            _ => Err(anyhow!("Unknown runtime: {}", container_runtime)),
+        }
+    }
 }
 
-pub fn set_namespace(ns_path: &str) -> Result<()> {
+pub fn set_namespace(ns_path: PathBuf) -> Result<()> {
     let fd: RawFd = File::open(ns_path)?.into_raw_fd();
     setns(fd, nix::sched::CloneFlags::CLONE_NEWNET)?;
     Ok(())
 }
 
-pub async fn get_container_namespace(container_id: String) -> Result<String> {
-    let channel = connect(CONTAINERD_SOCK_PATH).await?;
-    let mut client = ContainersClient::new(channel);
-    let request = GetContainerRequest { id: container_id };
-    let request = with_namespace!(request, DEFAULT_CONTAINERD_NAMESPACE);
-    let resp = client.get(request).await?;
-    let resp = resp.into_inner();
-    let container = resp
-        .container
-        .ok_or_else(|| anyhow!("container not found"))?;
-    let spec: Spec = serde_json::from_slice(
-        &container
-            .spec
-            .ok_or_else(|| anyhow!("invalid data from containerd"))?
-            .value,
-    )?;
-    let ns_path = spec
-        .linux
-        .namespaces
-        .iter()
-        .find(|ns| ns.ns_type == "network")
-        .ok_or_else(|| anyhow!("network namespace not found"))?
-        .path
-        .as_ref()
-        .ok_or_else(|| anyhow!("no network namespace path"))?;
-    Ok(ns_path.to_owned())
+async fn get_docker_container_pid(container_id: String) -> Result<PathBuf> {
+    let client = Docker::connect_with_local_defaults()?;
+    let inspect_options = Some(InspectContainerOptions { size: false });
+    let inspect_response = client
+        .inspect_container(&container_id, inspect_options)
+        .await?;
+
+    let pid = inspect_response
+        .state
+        .and_then(|state| state.pid)
+        .ok_or_else(|| anyhow!("No pid found"))?;
+
+    Ok(PathBuf::from(pid.to_string()))
```

Comment:
I would leave converting the `pid` into a `PathBuf` on the caller side, so here we return just the `pid` as a number, and when some other function needs to call like `set_namespace`, then they can do the conversion there like:

```rust
set_namespace(PathBuf::from(pid));
```

Reply:
the callee converts it to PathBuf because the `get_docker_container_pid` returns an i64 for the process however the containerd one returns a u32. I dont know if we should cast i64 to u32 but i think PIDs cant be negative. WDYT @aviramha